4.3-pipelines/demonstrators/4.3.1-genotyping/stepA-calculate-frequency.nf

```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env nextflow
+
+
+input_data_ch = Channel.fromPath(params.inputData)
+    .splitCsv(header: true, sep: "\t", strip: true)
+    .map { row -> [row.sample_id, row.fetch_mode, row.path] }
+
+final_vcf_name = file(params.outputVcf).getName()
+final_vcf_dir = file(params.outputVcf).getParent()
+bin_dir = file(params.binDir)
+
+
+process fetchReferenceGenome {
+
+    errorStrategy "retry"
+    maxRetries 3
+    publishDir params.debugDir
+
+    output:
+        path "reference.fa" into reference_genome_fa
+
+    """
+    wget -qO- ${params.referenceGenomeLink} | gzip -cd > reference.fa
+    """
+
+}
+
+
+process fetchInputData {
+
+    // samtools library cannot handle network errors, so we need to retry explicitly in case they happen.
+    errorStrategy "retry"
+    maxRetries 5
+    publishDir params.debugDir
+
+    input:
+        set sample_id, fetch_mode, path from input_data_ch
+    output:
+        set sample_id, file("${sample_id}.bam") into reads_bam
+
+    script:
+
+        if( fetch_mode == "EGA" )
+            """
+            pyega3 -d -t fetch "${path}" \
+                --max-retries 3 --retry-wait 10 \
+                -r 17 -s 61554422 -e 61575741 \
+                --format BAM --saveto "${sample_id}.bam"
+            """
+
+        else if ( fetch_mode == "FTP" )
+            """
+            ${bin_dir}/samtools/samtools view -b "${path}" chr17:63477061-63498373 > "${sample_id}.bam"
+            """
+
+        else
+            error "Unsupported data fetch mode: ${fetch_mode}"
+
+}
+
+
+process callVariants {
+
+    publishDir params.debugDir
+
+    input:
+        set sample_id, file("${sample_id}.bam") from reads_bam
+        file("reference.fa") from reference_genome_fa
+    output:
+        file("${sample_id}.vcf.gz") into calls_vcf
+        file("${sample_id}.vcf.gz.tbi") into indexes_tbi
+
+    """
+    bcftools mpileup -Ou -f "reference.fa" "${sample_id}.bam" \
```

Comment:
I think you need fullpath to `bcftools`, `bgzip` and `tabix`. There are couple of other places where the full path is not set either

Reply:
I think that specifying a full path would defeat the portability of the workflow? I would have all of the software in a docker/singularity container and then you would need to specify the full path.

Reply:
Right, so originally I had doubts about dockerising this stuff, because it limits us to only those platforms which can run Docker images in some form. I even tried to see a way around this and to bundle the required binaries together with the code, at least for this iteration of development, to speed things up.

However, I soon realised that any sort of portability is really impossible in the modern world without containerisation. So I went ahead and did it. Containers are now used throughout the pipeline, with no binaries and no full paths being present.